File: api/Enkl.Api/Services/OrganisationService.cs
using Enkl.Api.Auth;
using Enkl.Api.Data;
using Enkl.Api.Domain;
using Enkl.Api.Domain.Entities;
using Enkl.Api.Dtos;
using Enkl.Api.Validation;
using FluentValidation;
using Microsoft.EntityFrameworkCore;

namespace Enkl.Api.Services;

public class OrganisationService
{
    private readonly AppDbContext _db;
    private readonly IValidator<CreateUserRequest> _createUserValidator;
    private readonly SseBroadcaster _broadcaster;

    public OrganisationService(AppDbContext db, IValidator<CreateUserRequest> createUserValidator, SseBroadcaster broadcaster)
    {
        _db = db;
        _createUserValidator = createUserValidator;
        _broadcaster = broadcaster;
    }

    public async Task<OrganisationDetailDto?> GetOrganisationAsync(Guid organisationId)
    {
        var org = await _db.Organisations
            .AsNoTracking()
            .Include(o => o.Users)
            .FirstOrDefaultAsync(o => o.Id == organisationId);
        if (org is null) return null;

        var online = _broadcaster.GetOnlineUserIds();
        return new OrganisationDetailDto(
            org.Id, org.Name,
            org.DefaultNewUserPasswordHash is not null,
            org.Users.Select(u => new OrgUserDto(u.Id, u.Username, u.EmailAddress, u.DisplayName, u.IsOrgAdmin, u.IsActive, u.CreatedAt, online.Contains(u.Id))).ToList());
    }

    /// <summary>
    /// Lets an OrgAdmin configure the password newly (implicitly) created users in their org get,
    /// instead of the hardcoded PasswordHasher.GlobalDefaultNewUserPassword every org used to share.
    /// Only the bcrypt HASH is ever persisted (see Organisation.DefaultNewUserPasswordHash's own
    /// comment) — there is deliberately no corresponding "get the current default password" endpoint;
    /// an admin who forgets what they set can only overwrite it with a new one, never read it back.
    /// </summary>
    public async Task<bool> SetDefaultNewUserPasswordAsync(Guid organisationId, string password)
    {
        if (string.IsNullOrEmpty(password) || password.Length < 8)
        {
            throw new ApiValidationException("Password must be at least 8 characters.");
        }

        var org = await _db.Organisations.FirstOrDefaultAsync(o => o.Id == organisationId);
        if (org is null) return false;

        org.DefaultNewUserPasswordHash = PasswordHasher.Hash(password);
        await _db.SaveChangesAsync();
        return true;
    }

    /// <summary>The single place every implicit-account-creation path (MemberService.CreateAsync,
    /// MigrationEntityBuilder) resolves what a new User's PasswordHash should be — the org's own
    /// configured default if an OrgAdmin has set one, otherwise the system-wide fallback. Returns the
    /// HASH directly (never re-hashes an already-hashed value), same as CreateUserAsync's own
    /// PasswordHasher.Hash(request.Password) call one level up.</summary>
    public async Task<string> ResolveDefaultNewUserPasswordHashAsync(Guid organisationId)
    {
        var existingHash = await _db.Organisations
            .Where(o => o.Id == organisationId)
            .Select(o => o.DefaultNewUserPasswordHash)
            .FirstOrDefaultAsync();
        return existingHash ?? PasswordHasher.Hash(PasswordHasher.GlobalDefaultNewUserPassword);
    }

    /// <summary>Returns false if the target user doesn't exist or belongs to a different Organisation
    /// than the caller — an OrgAdmin can only manage users within their own org.</summary>
    public async Task<bool> SetUserAdminAsync(Guid callerOrganisationId, Guid targetUserId, bool isOrgAdmin)
    {
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Id == targetUserId);
        if (user is null || user.OrganisationId != callerOrganisationId) return false;

        user.IsOrgAdmin = isOrgAdmin;
        // Security review finding H2: without this, a demoted org-admin (or a newly-promoted one
        // whose token still carries the OLD orgAdmin=false claim) keeps using their existing token,
        // with its now-stale orgAdmin claim, until it naturally expires — up to 8 hours.
        user.SecurityStamp = Guid.NewGuid();
        await _db.SaveChangesAsync();
        return true;
    }

    /// <summary>
    /// Explicit account creation by an OrgAdmin, distinct from the implicit account-per-name creation
    /// MemberService/MigrationService do when adding a project member — here the admin sets a real
    /// username and an initial password directly (not the org's configured default/global fallback
    /// those other paths use — see ResolveDefaultNewUserPasswordHashAsync), and the new user is
    /// required to change it on first login, same as everywhere else a password gets set on someone's
    /// behalf. Usernames are unique across the whole system, not
    /// just this Organisation — matches how login (AuthController) resolves a username with no org
    /// scoping at all. Email is required here (unlike the implicit-creation paths, which can fall
    /// back to leaving it blank and flagging it for later) since an OrgAdmin explicitly filling out
    /// this form has no excuse not to supply one — it's the planned SAML2 identifier.
    /// </summary>
    public async Task<OrgUserDto> CreateUserAsync(Guid organisationId, CreateUserRequest request)
    {
        await _createUserValidator.ValidateAndThrowApiExceptionAsync(request);

        var displayName = (request.DisplayName ?? "").Trim();
        if (displayName.Length > 200) displayName = displayName[..200];

        var normalized = UsernameNormalizer.Normalize(request.Username ?? "");
        if (await _db.Users.AnyAsync(u => u.NormalizedUsername == normalized))
        {
            throw new ApiValidationException($"Username \"{normalized}\" is already taken.");
        }

        var (email, normalizedEmail) = await EmailValidation.ValidateAndNormalizeAsync(_db, request.EmailAddress, requireEmail: true, excludeUserId: null);

        var user = new User
        {
            Id = Guid.NewGuid(),
            OrganisationId = organisationId,
            Username = normalized,
            NormalizedUsername = normalized,
            EmailAddress = email,
            NormalizedEmailAddress = normalizedEmail,
            PasswordHash = PasswordHasher.Hash(request.Password),
            DisplayName = displayName,
            MustChangePassword = true,
            IsOrgAdmin = false,
            CreatedAt = DateTime.UtcNow
        };
        _db.Users.Add(user);
        await _db.SaveChangesAsync();

        return new OrgUserDto(user.Id, user.Username, user.EmailAddress, user.DisplayName, user.IsOrgAdmin, user.IsActive, user.CreatedAt, false);
    }

    /// <summary>
    /// The backfill path for a User created before this field existed (or migrated without one, see
    /// MigrationService's warnings) — same validation as CreateUserAsync, scoped to the caller's own
    /// Organisation the same way SetUserAdminAsync is.
    /// </summary>
    public async Task<bool> SetUserEmailAsync(Guid callerOrganisationId, Guid targetUserId, string emailAddress)
    {
        var user = await _db.Users.FirstOrDefaultAsync(u => u.Id == targetUserId);
        if (user is null || user.OrganisationId != callerOrganisationId) return false;

        var (email, normalizedEmail) = await EmailValidation.ValidateAndNormalizeAsync(_db, emailAddress, requireEmail: true, excludeUserId: user.Id);
        user.EmailAddress = email;
        user.NormalizedEmailAddress = normalizedEmail;
        await _db.SaveChangesAsync();
        return true;
    }

    /// <summary>Read-only listing for the SSO & Provisioning modal's Org Teams section — see
    /// OrgTeamSummaryDto's own comment for why there's no corresponding write method here.</summary>
    public async Task<List<OrgTeamSummaryDto>> GetOrgTeamsAsync(Guid organisationId)
    {
        var teams = await _db.OrgTeams
            .AsNoTracking()
            .Include(t => t.Members).ThenInclude(m => m.User)
            .Where(t => t.OrganisationId == organisationId)
            .OrderBy(t => t.Name)
            .ToListAsync();

        return teams.Select(t => new OrgTeamSummaryDto(
            t.Id, t.Name,
            t.Members.Select(m => new OrgTeamMemberSummaryDto(m.UserId, m.User.DisplayName)).ToList()
        )).ToList();
    }
}

File: php-api/src/Services/OrganisationService.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Services;

use Enkl\Api\Auth\PasswordHasher;
use Enkl\Api\Auth\UsernameNormalizer;
use Enkl\Api\Support\Uuid;
use Enkl\Api\Validation\ApiValidationException;
use PDO;

/** Ported from Services/OrganisationService.cs. */
final class OrganisationService
{
    public function __construct(private readonly PDO $db)
    {
    }

    public function getOrganisation(string $organisationId): ?array
    {
        $stmt = $this->db->prepare('SELECT "Id", "Name", "DefaultNewUserPasswordHash" FROM "Organisations" WHERE "Id" = :id');
        $stmt->execute(['id' => $organisationId]);
        $org = $stmt->fetch();
        if ($org === false) {
            return null;
        }

        $stmt = $this->db->prepare(
            'SELECT "Id", "Username", "EmailAddress", "DisplayName", "IsOrgAdmin", "IsActive", "CreatedAt" FROM "Users" WHERE "OrganisationId" = :id'
        );
        $stmt->execute(['id' => $organisationId]);
        $online = $this->onlineUserIds();
        $users = array_map(static fn(array $u): array => [
            'id' => $u['Id'],
            'username' => $u['Username'],
            'emailAddress' => $u['EmailAddress'],
            'displayName' => $u['DisplayName'],
            'isOrgAdmin' => (bool) $u['IsOrgAdmin'],
            'isActive' => (bool) $u['IsActive'],
            'createdAt' => $u['CreatedAt'],
            'isOnline' => in_array($u['Id'], $online, true),
        ], $stmt->fetchAll());

        return [
            'id' => $org['Id'],
            'name' => $org['Name'],
            'hasCustomDefaultPassword' => $org['DefaultNewUserPasswordHash'] !== null,
            'users' => $users,
        ];
    }

    /**
     * Lets an OrgAdmin configure the password newly (implicitly) created users in their org get,
     * instead of the hardcoded PasswordHasher::GLOBAL_DEFAULT_NEW_USER_PASSWORD every org used to
     * share. Only the bcrypt HASH is ever persisted — there is deliberately no corresponding "get the
     * current default password" endpoint; an admin who forgets what they set can only overwrite it
     * with a new one, never read it back.
     */
    public function setDefaultNewUserPassword(string $organisationId, string $password): bool
    {
        if (strlen($password) < 8) {
            throw new ApiValidationException('Password must be at least 8 characters.');
        }

        $stmt = $this->db->prepare('SELECT 1 FROM "Organisations" WHERE "Id" = :id');
        $stmt->execute(['id' => $organisationId]);
        if ($stmt->fetch() === false) {
            return false;
        }

        $stmt = $this->db->prepare('UPDATE "Organisations" SET "DefaultNewUserPasswordHash" = :hash WHERE "Id" = :id');
        $stmt->execute(['hash' => PasswordHasher::hash($password), 'id' => $organisationId]);
        return true;
    }

    /** @return string[] Same query/grace-window as ChatService::onlineUserIds — duplicated rather
     * than shared, matching this tier's existing per-class-duplication convention (see php-api/CLAUDE.md). */
    private function onlineUserIds(): array
    {
        $stmt = $this->db->query('SELECT "UserId" FROM "SsePresence" WHERE "LastSeenAt" > now() - interval \'25 seconds\'');
        return array_column($stmt->fetchAll(), 'UserId');
    }

    /** Returns false if the target user doesn't exist or belongs to a different Organisation than the caller. */
    public function setUserAdmin(string $callerOrganisationId, string $targetUserId, bool $isOrgAdmin): bool
    {
        $stmt = $this->db->prepare('SELECT "OrganisationId" FROM "Users" WHERE "Id" = :id');
        $stmt->execute(['id' => $targetUserId]);
        $row = $stmt->fetch();
        if ($row === false || $row['OrganisationId'] !== $callerOrganisationId) {
            return false;
        }

        // Security review finding H2: rotating SecurityStamp here invalidates the target's
        // already-issued token(s), whose orgAdmin claim would otherwise stay stale (still
        // false/true from before this change) for up to the token's full 8-hour lifetime.
        $stmt = $this->db->prepare('UPDATE "Users" SET "IsOrgAdmin" = :admin, "SecurityStamp" = gen_random_uuid() WHERE "Id" = :id');
        // (int), not the raw PHP bool — PDO's array-form execute() would bind false as '' otherwise,
        // which Postgres's boolean parser rejects.
        $stmt->execute(['admin' => (int) $isOrgAdmin, 'id' => $targetUserId]);
        return true;
    }

    /**
     * Explicit account creation by an OrgAdmin, distinct from the implicit account-per-name creation
     * MemberService/MigrationService do when adding a project member — here the admin sets a real
     * username and initial password directly, and the new user must change it on first login.
     * Usernames are unique across the whole system, not just this Organisation. Email is required
     * here (unlike the implicit-creation paths, which can leave it blank and flag it for later)
     * since an OrgAdmin filling out this form explicitly has no excuse not to supply one — it's the
     * planned SAML2 identifier.
     */
    public function createUser(string $organisationId, array $request): array
    {
        $displayName = trim((string) ($request['displayName'] ?? ''));
        if ($displayName === '') {
            throw new ApiValidationException('Please enter a display name.');
        }
        if (strlen($displayName) > 200) {
            $displayName = substr($displayName, 0, 200);
        }

        $password = (string) ($request['password'] ?? '');
        if (strlen($password) < 8) {
            throw new ApiValidationException('Password must be at least 8 characters.');
        }

        $normalized = UsernameNormalizer::normalize((string) ($request['username'] ?? ''));
        if ($normalized === '') {
            throw new ApiValidationException('Please enter a username.');
        }

        $stmt = $this->db->prepare('SELECT 1 FROM "Users" WHERE "NormalizedUsername" = :n');
        $stmt->execute(['n' => $normalized]);
        if ($stmt->fetch() !== false) {
            throw new ApiValidationException("Username \"{$normalized}\" is already taken.");
        }

        [$email, $normalizedEmail] = EmailValidation::validateAndNormalize($this->db, $request['emailAddress'] ?? null, true, null);

        $id = Uuid::v4();
        $stmt = $this->db->prepare(<<<SQL
            INSERT INTO "Users" ("Id", "OrganisationId", "Username", "NormalizedUsername", "EmailAddress", "NormalizedEmailAddress", "PasswordHash", "DisplayName", "MustChangePassword", "IsOrgAdmin", "CreatedAt")
            VALUES (:id, :orgId, :username, :normalized, :email, :normalizedEmail, :hash, :displayName, true, false, now())
        SQL);
        $stmt->execute([
            'id' => $id,
            'orgId' => $organisationId,
            'username' => $normalized,
            'normalized' => $normalized,
            'email' => $email,
            'normalizedEmail' => $normalizedEmail,
            'hash' => PasswordHasher::hash($password),
            'displayName' => $displayName,
        ]);

        return [
            'id' => $id,
            'username' => $normalized,
            'emailAddress' => $email,
            'displayName' => $displayName,
            'isOrgAdmin' => false,
            'isActive' => true,
            'createdAt' => gmdate('Y-m-d\TH:i:s.v\Z'),
        ];
    }

    /**
     * The backfill path for a User created before this field existed (or migrated without one, see
     * MigrationService's warnings) — same validation as createUser, scoped to the caller's own
     * Organisation the same way setUserAdmin is.
     */
    public function setUserEmail(string $callerOrganisationId, string $targetUserId, ?string $emailAddress): bool
    {
        $stmt = $this->db->prepare('SELECT "OrganisationId" FROM "Users" WHERE "Id" = :id');
        $stmt->execute(['id' => $targetUserId]);
        $row = $stmt->fetch();
        if ($row === false || $row['OrganisationId'] !== $callerOrganisationId) {
            return false;
        }

        [$email, $normalizedEmail] = EmailValidation::validateAndNormalize($this->db, $emailAddress, true, $targetUserId);

        $stmt = $this->db->prepare('UPDATE "Users" SET "EmailAddress" = :email, "NormalizedEmailAddress" = :normalizedEmail WHERE "Id" = :id');
        $stmt->execute(['email' => $email, 'normalizedEmail' => $normalizedEmail, 'id' => $targetUserId]);
        return true;
    }

    /** Read-only listing for the SSO & Provisioning modal's Org Teams section — SCIM (ScimGroupService)
     * is the only writer of OrgTeams/OrgTeamMember, mirroring GetOrgTeamsAsync in OrganisationService.cs. */
    public function getOrgTeams(string $organisationId): array
    {
        $stmt = $this->db->prepare('SELECT "Id", "Name" FROM "OrgTeams" WHERE "OrganisationId" = :id ORDER BY "Name"');
        $stmt->execute(['id' => $organisationId]);
        $teams = $stmt->fetchAll();

        $memberStmt = $this->db->prepare(<<<SQL
            SELECT m."UserId", u."DisplayName" FROM "OrgTeamMember" m
            JOIN "Users" u ON u."Id" = m."UserId"
            WHERE m."OrgTeamId" = :id
        SQL);

        return array_map(function (array $t) use ($memberStmt): array {
            $memberStmt->execute(['id' => $t['Id']]);
            return [
                'id' => $t['Id'],
                'name' => $t['Name'],
                'members' => array_map(
                    static fn(array $m): array => ['userId' => $m['UserId'], 'displayName' => $m['DisplayName']],
                    $memberStmt->fetchAll()
                ),
            ];
        }, $teams);
    }
}
